### PR TITLE
feat(website): fix issues with create website with ftp failed

### PR DESCRIPTION
**Pull request by zhengkunwang223**

### CODE REVIEW

File: agent/app/service/website.go [hunk 1]
```diff
@@ -2130,7 +2123,7 @@ func (w WebsiteService) UpdatePathAuthBasic(req request.NginxPathAuthUpdate) err
 	directives := config.Directives
 	location, _ := directives[0].(*components.Location)
 	location.UpdateDirective("auth_basic_user_file", []string{fmt.Sprintf("/www/sites/%s/path_auth/pass/%s", website.Alias, fmt.Sprintf("%s.pass", req.Name))})
-	location.ChangePath("~*", req.Path)
+	location.ChangePath("~*", fmt.Sprintf("^%s", req.Path))
 	var passwdHash []byte
 	passwdHash, err = bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
 	if err != nil {
```
Comment (wanghe-fit2cloud) [inline agent/app/service/website.go:2129]: The provided code has some areas that can be reviewed for better organization, clarity, and performance.

### Code Snippet 1

```go
if len(create.FtpUser) != 0 && len(create.FtpPassword) != 0 {
    createFtpUser := func(t *task.Task) error {
        indexDir := GetSitePath(*website, SiteIndexDir)
        itemID, err := NewIFtpService().Create(dto.FtpCreate{User: create.FtpUser, Password: create.FtpPassword, Path: indexDir})
        if err != nil {
            t.Log(fmt.Sprintf("create ftp for website failed, err: %v", err))
        }
        website.FtpID = itemID
        return nil
    }

    deleteFtpUser := func(t *task.Task) {
        if website.FtpID > 0 {
            if err = NewIFtpService().Delete(dto.BatchDeleteReq{Ids: []uint{website.FtpID}}); err != nil {
                t.Log(err.Error())
            }
        }
    }

    createTask.AddSubTask(i18n.GetWithName("ConfigFTP", create.FtpUser), createFtpUser, deleteFtpUser)
}
```

**Recommendations:**

- **Function Simplification:** Consider combining `createFtpUser` and `deleteFtpUser` into a single function that handles both operations simultaneously.
  
```go
func ensureFtpUserExists(t context.Context) func(ctx context.Context) error {
    return func(ctx context.Context) error {
        user := dto.FtpCreate{
            User:   create.FtpUser,
            Password: create.FtpPassword,
            Path:     GetSitePath(*website, SiteIndexDir),
        }
        
        itemID, err := NewIFtpService().Create(user)
        if err != nil {
            return err
        }
        
        website.FtpID = itemID
        
        // Cleanup on failure
        defer cleanupFunc(itemID)

        return nil
    }
}

func cleanupFunc(id uint) {
    if id > 0 {
        if err := NewIFtpService().Delete(dto.BatchDeleteReq{Ids: []uint{id}}); err != nil {
            log.Printf("Unable to clean up FTP resource: %v\n", err)
        }
    }
}
```

- **Context Use:** It's generally recommended to use contexts instead of task execution directly inside tasks. Ensure all functions used accept and handle contexts properly.

2. **Log Formatting:** Improve logging formatting to ensure consistency across different environments (`t.Log()` should be replaced with consistent logging methods according to specific project standards).

### Code Snippet 2

```go
if len(create.FtpUser) != 0 && len(create.FtpPassword) != 0 {
    createFtpUser := func(t *task.Task) error {
        ...
    }

    createTask.AddSubTask(i18n.GetWithName("ConfigFTP", create.FtpUser), createFtpUser)
}
```
No significant changes here but the pattern is similar to snippet 1.

## General Recommendations

1. **Consistent Structure**: Encourage adhering to a standard structure like one of the popular Go coding patterns such as interfaces, structs, or modules.

2. **Unit Test Coverage:** Always test edge cases including empty strings, null values, etc., especially where new code introduces logic flow.

3. **Code readability**: Avoid deeply nested blocks unless absolutely necessary. Refactor long functions/functions with many parameters if they make it hard to understand at glance.

4. **Optimization Tips**:
   - Minimize repeated calls to expensive operation like database queries within loops unless profiling indicates otherwise.
   - Utilize buffered channels or go-routines efficiently to perform parallel operations wherever feasible and safe.

5. **Avoid Global Variables**: If you need shared data between functions, consider using package-level variables cautiously after careful analysis to avoid race conditions.

6. **Linter Checks**: Regularly run linters (like golint, gosec) to identify and fix style inconsistencies before submitting for review.

**Overall Feedback**: The code seems well-structured except for minor inefficiencies and inconsistent logging practices which could impact maintainability and debugging effort. Suggestions focus on improving code cleanliness, consistency, and efficiency.

File: agent/app/service/runtime_utils.go [hunk 2]
```diff
@@ -331,20 +331,25 @@ func buildRuntime(runtime *model.Runtime, oldImageID string, oldEnv string, rebu
...
 		}
+		if out, err := compose.DownAndUp(composePath); err != nil {
+			runtime.Status = constant.StatusStartErr
+			runtime.Message = out
+			_ = runtimeRepo.Save(runtime)
+		}
+		runtime.Status = constant.StatusRunning
 	}
 	_ = runtimeRepo.Save(runtime)
 }
```
Comment (wanghe-fit2cloud) [inline agent/app/service/runtime_utils.go:355]: There is an issue with the logic inside the `if` block before saving the runtime status to running if no errors occur up to that point. Specifically:

1. The function should not call `_runtimeRepo.Save(runtime)` inside both branches of the conditional statement (before and after the installation command) since it may overwrite changes made in one branch.

2. After processing the installation command, there might be some additional cleanup needed (depending on what `compose.DownAndUp()` does), but this part needs clarification based on the actual implementation.

For optimization, consider moving any operations unrelated to runtime state updates above the conditional statements to reduce redundancy. Additionally, ensure error handling when calling `downAndUp()` since downgrading and upgrading the container can fail as well.

File: frontend/src/views/website/website/config/basic/auth-basic/index.vue [hunk 3]
```diff
@@ -207,7 +214,14 @@ const searchAll = () => {
...
+    getWebsite(id.value).then(async (res) => {
+        showPath.value = res.data.type !== 'proxy' && res.data.type !== 'deployment';
+    });
+};
+
 onMounted(() => {
+    getSiteDetail();
     searchAll();
 });
 </script>
```
Comment (wanghe-fit2cloud) [inline frontend/src/views/website/website/config/basic/auth-basic/index.vue:227]: There are several areas where this code can be improved:

### 1. Remove Unused Code

- The `createRef` is not used anywhere after its initialization.

```typescript
// remove this line: const createRef = ref();

```

Additionally, the unused methods `operateAuthConfig`, `getAuthConfig`, and possibly others might need to be removed as well if they are not referenced elsewhere.

### 2. Simplify Variable Declarations

Declare variables more clearly when using `ref()` or even directly assigning values during setup.

```typescript
const data = ref([]); // instead of let data = [];

const loading = ref(false);

...

onMounted(() => {
    loadData();
});
```

### 3. Refactor Complex Logic

If the complex logic related to checking conditions should only run under certain circumstances, consider wrapping it in a separate function with conditional checks within.

```javascript
function handleShowPath(websiteType) {
  showPath.value = websiteType !== 'proxy' && websiteType !== 'deployment';
}

...
handleShowPath(siteDetail.type);
```

This makes the code cleaner and easier to maintain.

### 4. Add More TypeScript Type Checking

Ensure that types are correctly defined and used throughout the component.

```typescript
interface WebSiteType { /** Define your interface here */ }

const siteDetail: WebSiteType = computed(() => {/* logic to compute siteDetail*/ });

... other similar uses.
```

### 5. Ensure Error Handling

Include error handling for APIs like `getWebsite` and `getPathData`.

```typescript
async function loadWebsites() {
   try {
      const response = await getWebsite(id.value);
      pathData.value = response.data;
   } catch (error) {
      console.error("Error fetching websites", error.message);
   }
}
loadWebsites();
```

These recommendations will help streamline the codebase by removing unnecessary parts, simplifying variable declarations, improving modularity, adding type definitions explicitly, and ensuring robustness through proper error handling.
